samples/grpc/local-drone-control-scala/src/main/scala/charging/ChargingStation.scala

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+package charging
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.TimerScheduler
+import akka.persistence.typed.RecoveryCompleted
+import akka.persistence.typed.ReplicaId
+import akka.persistence.typed.scaladsl.Effect
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import akka.persistence.typed.scaladsl.ReplicationContext
+import akka.projection.grpc.replication.scaladsl.ReplicatedBehaviors
+import akka.projection.grpc.replication.scaladsl.Replication
+import akka.projection.grpc.replication.scaladsl.Replication.EdgeReplication
+import akka.projection.grpc.replication.scaladsl.ReplicationSettings
+import akka.projection.r2dbc.scaladsl.R2dbcReplication
+import akka.serialization.jackson.CborSerializable
+
+import java.time.Instant
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.duration.DurationLong
+import scala.concurrent.duration.FiniteDuration
+import scala.math.Ordered.orderingToOrdered
+
+object ChargingStation {
+
+  // commands and replies
+  sealed trait Command extends CborSerializable {}
+  case class Create(
+      locationId: String,
+      chargingSlots: Int,
+      replyTo: ActorRef[Done])
+      extends Command
+  case class StartCharging(
+      droneId: String,
+      replyTo: ActorRef[StartChargingResponse])
+      extends Command
+  sealed trait StartChargingResponse
+  case class AllSlotsBusy(firstSlotFreeAt: Instant)
+      extends StartChargingResponse
+
+  case class GetState(replyTo: ActorRef[ChargingStation.State]) extends Command
+
+  private case class CompleteCharging(droneId: String) extends Command
+
+  // events
+  sealed trait Event extends CborSerializable {}
+  case class Created(locationId: String, chargingSlots: Int) extends Event
+  case class ChargingStarted(droneId: String, chargeComplete: Instant)
+      extends Event
+      with StartChargingResponse
+
+  case class ChargingCompleted(droneId: String) extends Event
+
+  case class ChargingDrone(
+      droneId: String,
+      chargingDone: Instant,
+      replicaId: String)
+  case class State(
+      chargingSlots: Int,
+      dronesCharging: Set[ChargingDrone],
+      stationLocationId: String)
+      extends CborSerializable
+
+  val EntityType = "charging-station"
+
+  private val FullChargeTime = 5.minutes
+
+  // FIXME This is the only difference from the cloud one, maybe include in both and keep the RES identical?
```

Comment:
I later realised that is not quite possible, since edge needs a different init method in Replication.grpcEdgeReplication and adding consumer filters, so I now added two separate methods (but duplicated the full sources with both those in both cloud and edge services).